doc/api-and-naming.md

```diff
@@ -0,0 +1,94 @@
+# API and naming
+
+## General API conventions
+
+Most of the functions in bash-completion generate completions and directly
+inject them to the `COMPREPLY` array variable, as required for completions to
+work.
+
+Most other functions make use of "output" variables, i.e. assign values to
+them. The most common one of these is named `ret`. Consult the commentary
+before each function in the source to find out the specific names.
+`local`izing output variables before invoking a function that populates them
+is the caller's responsibility.
+Note that if calling multiple functions that assign output to the same variable
+during one completion function run, each result should be copied to another
+variable between the calls to avoid it possibly being overwritten and lost on
+the next call. Also, the variables should also be ensured to be clear before
+each call, typically by `unset -v`ing them when multiple such calls are used,
+to avoid them interfering with each other.
```

Comment:
Do we always require to clear the state by `unset -v`? I feel we can require it only when the function called next will use the state of the variable (i.e., the value, the existence of the value, and/or the existence of the variable). Otherwise, we always need to save the values including the following case:

```bash
# I would like to allow writing this way as far as
# _func1/_func2 doesn't reference the original state of ret.
local ret
_func1 "$1"
_func2 "$ret"

# This might be much safer but feels redundant
local ret
unset -v ret
_func1 "$1"
local intermediate=$ret
unset -v ret
_func2 "$intermediate"
```

Reply:
Agreed, addressed in 1fb425f827629f7419433bf1e6a84cda5fe38897 I hope.

Reply:
Thanks!